apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts

```diff
@@ -0,0 +1,362 @@
+  ) {}
+
+  // We initialise required values
+  private init(user: UserWithRole, proposalPk: number, filter?: string) {
+    if (user && proposalPk) {
+      this.userWithRole = user;
+      this.proposalPk = proposalPk;
+      this.proposalFilter = filter ?? null;
+    } else {
+      throw new Error(
+        'Can not download proposal pdf , request missing required information'
+      );
+    }
+  }
+
+  private async getProposal() {
+    let proposal;
+    if (this.proposalFilter && this.proposalFilter === 'id') {
+      proposal = await this.proposalDataSource.getProposalById(
+        this.proposalPk.toString()
+      );
+    } else {
+      proposal = await this.proposalDataSource.get(this.proposalPk);
+    }
+
+    if (proposal === null) {
+      throw new Error('Proposal not found');
+    }
+
+    return proposal;
+  }
+
+  private async getPdfTemplate(): Promise<PdfTemplate | null> {
+    const callId = this.proposalPDFData?.proposal?.callId;
+    if (callId) {
+      const proposalCall = await this.callDataSource.getCall(callId);
+      if (proposalCall === null) {
+        throw new Error('Proposal call not be found');
+      }
+      if (proposalCall.pdfTemplateId) {
+        return (
+          await this.pdfTemplateDataSource.getPdfTemplates({
+            filter: {
+              templateIds: [proposalCall.pdfTemplateId],
+            },
+          })
+        )[0];
+      }
+    }
+
+    return null;
+  }
+
+  private async getprincipalInvestigator(): Promise<BasicUserDetails> {
+    let principalInvestigator = null;
+    const proposerId = this.proposalPDFData.proposal?.proposerId;
+
+    if (proposerId) {
+      principalInvestigator = await this.userDataSource.getBasicUserInfo(
+        proposerId
+      );
+    }
+
+    if (!principalInvestigator) {
+      throw new Error('Proposal principal investigator not found');
+    }
+
+    return principalInvestigator;
+  }
+  private async getCoProposers(): Promise<BasicUserDetails[] | null> {
+    let coProposers = null;
+    const proposalPk = this.proposalPDFData.proposal?.primaryKey;
+    if (proposalPk) {
+      coProposers = await this.userDataSource.getProposalUsers(proposalPk);
+    }
+
+    return coProposers;
+  }
+
+  private async getQuestionary(questionaryId: number) {
+    const questionary = await this.questionaryDataSource.getQuestionary(
+      questionaryId
+    );
+
+    if (!questionary) {
+      throw new Error(`Questionary with ID '${questionaryId}' not found`);
+    }
+
+    return questionary;
+  }
+
+  private async getQuestionarySteps(): Promise<QuestionaryStep[]> {
+    let questionarySteps = null;
+    const questionaryId = this.proposalPDFData.proposal?.questionaryId;
+    if (questionaryId) {
+      questionarySteps = await this.questionaryDataSource.getQuestionarySteps(
+        questionaryId
+      );
+    }
+    if (isRejection(questionarySteps) || questionarySteps == null) {
+      throw new Error('Could not fetch questionary');
```

Comment:
Should we add some logging here? Looks like there is a `rejection.reason` has a piece of valuable information that should be stored or returned